canary/webrtc-c/src/CanaryWebrtcStorage.cpp

```diff
@@ -0,0 +1,380 @@
+#include "Include.h"
+
+STATUS onNewConnection(Canary::PPeer);
+STATUS run(Canary::PConfig);
+VOID runPeer(Canary::PConfig, TIMER_QUEUE_HANDLE, STATUS*);
+VOID sendLocalFrames(Canary::PPeer, MEDIA_STREAM_TRACK_KIND, const std::string&, UINT64, UINT32);
+VOID sendCustomFrames(Canary::PPeer, MEDIA_STREAM_TRACK_KIND, UINT64, UINT64);
+STATUS canaryRtpOutboundStats(UINT32, UINT64, UINT64);
+STATUS canaryKvsStats(UINT32, UINT64, UINT64);
+
+std::atomic<bool> terminated;
+VOID handleSignal(INT32 signal)
+{
+    UNUSED_PARAM(signal);
+    terminated = TRUE;
+}
+
+// add frame pts, original frame size, CRC to beginning of buffer after Annex-B format NALu
+VOID addCanaryMetadataToFrameData(PBYTE buffer, PFrame pFrame)
+{
+    PBYTE pCurPtr = buffer + ANNEX_B_NALU_SIZE;
+    putUnalignedInt64BigEndian((PINT64) pCurPtr, pFrame->presentationTs);
+    pCurPtr += SIZEOF(UINT64);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, pFrame->size);
+    pCurPtr += SIZEOF(UINT32);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, COMPUTE_CRC32(buffer, pFrame->size));
+}
+
+// Frame Data format: NALu (4 bytes) + Header (PTS, Size (including header), CRC (frame data) + Randomly generated frameBits
+
+// TODO: Support dynamic random frame sizes to bring it closer to real time scenarios
+VOID createCanaryFrameData(PBYTE buffer, PFrame pFrame)
+{
+    UINT32 i;
+    // For decoding purposes, the first 4 bytes need to be a NALu
+    putUnalignedInt32BigEndian((PINT32) buffer, 0x00000001);
+    for (i = ANNEX_B_NALU_SIZE + CANARY_METADATA_SIZE; i < pFrame->size; i++) {
+        buffer[i] = RAND();
+    }
+    addCanaryMetadataToFrameData(buffer, pFrame);
+}
+
+INT32 main(INT32 argc, CHAR* argv[])
+{
+#ifndef _WIN32
+    signal(SIGINT, handleSignal);
+#endif
+
+    STATUS retStatus = STATUS_SUCCESS;
+    initializeEndianness();
+    SET_INSTRUMENTED_ALLOCATORS();
+    SRAND(time(0));
+    // Make sure that all destructors have been called first before resetting the instrumented allocators
+    CHK_STATUS([&]() -> STATUS {
+        STATUS retStatus = STATUS_SUCCESS;
+        auto config = Canary::Config();
+
+        Aws::SDKOptions options;
+        Aws::InitAPI(options);
+
+        CHK_STATUS(config.init(argc, argv));
+        CHK_STATUS(run(&config));
+
+    CleanUp:
+
+        Aws::ShutdownAPI(options);
+
+        return retStatus;
+    }());
+
+CleanUp:
+
+    if (STATUS_FAILED(RESET_INSTRUMENTED_ALLOCATORS())) {
+        DLOGE("FOUND MEMORY LEAK");
```

Comment:
Just checking - does this cause the test to fail?

Reply:
Yes, causes it to fail